### PR TITLE
Fix ipfs address generation

### PR DESCRIPTION
**Pull request by adonesky1**
- CHANGED:

  -**BREAKING** Change IPFS URL generation to use subdomains and [cidV1s over cidV0s](https://docs.ipfs.io/concepts/content-addressing/#identifier-formats), in order to enhance [origin based security](https://docs.ipfs.io/concepts/ipfs-gateway/#subdomain) in our use of IPFS assets
   - Consumers using IPFS Gateway which do not support subdomain formats will need to set the new config value 'useIPFSSubdomains' on CollectiblesController to false in order to have continued IPFS resolution support.

### CODE REVIEW

**Review by wachunei — DISMISSED** [review]
LGTM

File: src/util.ts [hunk 1]
```diff
@@ -769,19 +770,81 @@ export function validateMinimumIncrease(proposed: string, min: string) {
...
+export function getIpfsCIDv1AndPath(
+  ipfsUrl: string,
+): { cid: string; path?: string } {
+  const url = removeIpfsProtocolPrefix(ipfsUrl);
+
+  // check if there is a path
+  // (CID is everything preceding first forward slash, path is everything after)
+  const index = url.indexOf('/');
+  const cid = index !== -1 ? url.substring(0, index) : url;
+  const path = index !== -1 ? url.substring(index) : '';
```
Comment (Gudahtt) [inline src/util.ts:805]: Nit: Could we default to `undefined` here for `path` instead, if it's not present? That would match the type signature for this function.

File: src/util.ts [hunk 2]
```diff
@@ -769,19 +770,81 @@ export function validateMinimumIncrease(proposed: string, min: string) {
...
+ */
+export function getFormattedIpfsUrl(
+  ipfsGateway: string,
+  ipfsUrl: string,
+  subdomainSupported: boolean,
+): string {
+  if (subdomainSupported) {
+    const gatewayHost = new URL(addUrlProtocolPrefix(ipfsGateway)).host;
+    const { cid, path } = getIpfsCIDv1AndPath(ipfsUrl);
+    return `https://${cid}.ipfs.${gatewayHost}${path}`;
```
Comment (Gudahtt) [inline src/util.ts:843]: It would be better to extract the protocol from the IPFS gateway as well, if it's specified. Users might want to use alternative protocols.

```suggestion
    const { host, protocol } = new URL(addUrlProtocolPrefix(ipfsGateway));
    const { cid, path } = getIpfsCIDv1AndPath(ipfsUrl);
    return `${protocol}//${cid}.ipfs.${host}${path}`;
```

File: src/util.ts [hunk 3]
```diff
@@ -769,19 +770,81 @@ export function validateMinimumIncrease(proposed: string, min: string) {
...
+    return `https://${urlString}`;
   }
+  return urlString;
+}
 
-  return url;
+/**
+ * Formats url correctly for use retrieving assets hosted on IPFS.
+ *
+ * @param ipfsGateway - the user preferred ipfsGateway.
```
Comment (Gudahtt) [inline src/util.ts:830]: Nit: These descriptions could be better. Keep in mind that these aren't just code comments, they are inline documentation. This gets used to build our API documentation (which isn't automatically built/hosted right now, but it will be soon, and it can be generated manually). Better to use proper case (start the sentence with a capital letter, capitalize acronyms like IPFS, use words and not variable names, etc.), and to include relevant details like what sort of string this is.

For example, it looks like this is expected to be the URL for the IPFS gateway, except that the protocol is optional, and the trailing `/ipfs` path is optional.

The same general point applies to most of the other inline docs as well.

File: src/util.ts [hunk 4]
```diff
@@ -769,19 +770,81 @@ export function validateMinimumIncrease(proposed: string, min: string) {
...
+): { cid: string; path?: string } {
+  const url = removeIpfsProtocolPrefix(ipfsUrl);
+
+  // check if there is a path
+  // (CID is everything preceding first forward slash, path is everything after)
+  const index = url.indexOf('/');
+  const cid = index !== -1 ? url.substring(0, index) : url;
+  const path = index !== -1 ? url.substring(index) : '';
+
+  // we want to ensure that the CID is v1 (https://docs.ipfs.io/concepts/content-addressing/#identifier-formats)
```
Comment (Gudahtt) [inline src/util.ts:807]: Nit: Maybe worth mentioning why this is important to us specifically: because v0 isn't compatible with subdomain IPFS gateway URLs, but v1 is.

File: src/util.ts [hunk 5]
```diff
@@ -769,19 +770,81 @@ export function validateMinimumIncrease(proposed: string, min: string) {
...
+  ipfsUrl: string,
+  subdomainSupported: boolean,
+): string {
+  if (subdomainSupported) {
+    const gatewayHost = new URL(addUrlProtocolPrefix(ipfsGateway)).host;
+    const { cid, path } = getIpfsCIDv1AndPath(ipfsUrl);
+    return `https://${cid}.ipfs.${gatewayHost}${path}`;
+  }
+  const cidAndPath = removeIpfsProtocolPrefix(ipfsUrl);
+  const gateway = ipfsGateway.endsWith('/ipfs/')
```
Comment (Gudahtt) [inline src/util.ts:846]: The protocol gets added in the subdomain branch (via `addUrlProtocolPrefix`), but it's not added here. So this function now returns an invalid URL in some cases, if `subdomainSupported` is false and the IPFS gateway passed in doesn't include a protocol.

Also, is there a reason we're looking for this path in particular, and with a trailing slash? We could make this more robust by always ignoring the path. I guess it's not super important though, if this is just to avoid mobile breaking changes.

File: src/util.ts [hunk 6]
```diff
@@ -769,19 +770,81 @@ export function validateMinimumIncrease(proposed: string, min: string) {
...
+  };
+}
+
+/**
+ * Adds URL protocol prefix to url string if missing.
+ *
+ * @param urlString - Ipfs url.
+ * @returns string.
+ */
+export function addUrlProtocolPrefix(urlString: string): string {
```
Comment (Gudahtt) [inline src/util.ts:820]: This appears to have no tests

**Comment by Gudahtt** [comment]
Overall this looks good! I left some suggestions for improvements

File: src/util.ts (hunk 4 above)
Comment (adonesky1) [reply src/util.ts:807]: Agreed, but as a side note, apparently this isn't always the case. Found some instances where v0 CIDS worked in the subdomain format. Will add to this note anyways!

File: src/util.ts (hunk 6 above)
Comment (adonesky1) [reply src/util.ts:821]: added in https://github.com/MetaMask/controllers/pull/655/commits/d4c8ade79f415cfac20002ae2cc47780f0b44e7a

File: src/util.ts (hunk 5 above)
Comment (adonesky1) [reply src/util.ts]: both comments addressed here: https://github.com/MetaMask/controllers/pull/655/commits/d4c8ade79f415cfac20002ae2cc47780f0b44e7a

File: src/util.ts (hunk 4 above)
Comment (adonesky1) [reply src/util.ts]: added to his comment here: https://github.com/MetaMask/controllers/pull/655/commits/3432359baba873cea2947164581603277438f990

File: src/util.ts (hunk 2 above)
Comment (adonesky1) [reply src/util.ts]: done here: https://github.com/MetaMask/controllers/pull/655/commits/d4c8ade79f415cfac20002ae2cc47780f0b44e7a

File: src/util.ts (hunk 1 above)
Comment (adonesky1) [reply src/util.ts]: done here: https://github.com/MetaMask/controllers/pull/655/commits/d4c8ade79f415cfac20002ae2cc47780f0b44e7a

**Review by wachunei — APPROVED** [review]

File: src/util.ts (hunk 4 above)
Comment (Gudahtt) [reply src/util.ts]: Really? Since v0 is mixed-case, it seemed unlikely that any would be completely lower case. But I guess that's possible.

File: src/util.ts [hunk 7]
```diff
@@ -769,19 +770,79 @@ export function validateMinimumIncrease(proposed: string, min: string) {
...
+}
+
+/**
+ * Adds URL protocol prefix to input URL string if missing.
+ *
+ * @param urlString - An IPFS URL.
+ * @returns A URL with a https:// prepended.
+ */
+export function addUrlProtocolPrefix(urlString: string): string {
+  if (!urlString.match(/(^http:\/\/)|(^https:\/\/)/u)) {
```
Comment (Gudahtt) [inline src/util.ts:822]: NIt: This still only allows `http` and `https` protocols, so the protocol support added below only helps with HTTP support. We could instead check here to ensure _any_ protocol is set.

This might be an extreme edge case though. I don't know of any IPFS gateways run on non-HTTP/HTTPS protocols :thinking:  I don't know much about this space though.

File: src/util.test.ts [hunk 8]
```diff
@@ -1068,23 +1073,129 @@ describe('util', () => {
...
+  });
+
+  describe('addUrlProtocolPrefix', () => {
+    it('should return a URL with https:// prepended if input URL does not already have it', () => {
+      expect(util.addUrlProtocolPrefix(IFPS_GATEWAY)).toStrictEqual(
+        `https://${IFPS_GATEWAY}`,
       );
     });
+
+    it('should return a URL as is if https:// is already prepended', () => {
```
Comment (Gudahtt) [inline src/util.test.ts:1167]: Nit: It's probably also worth ensuring that `http` URLs are handled correctly as well

**Review by Gudahtt — APPROVED** [review]
LGTM!

File: src/util.test.ts (hunk 8 above)
Comment (adonesky1) [reply src/util.test.ts:1167]: Will handle in follow up PR!

File: src/util.ts (hunk 7 above)
Comment (adonesky1) [reply src/util.ts:822]: Will research and handle in follow up PR!
